From e556fb5b4d366200345618fd6326440e4b59bfc4 Mon Sep 17 00:00:00 2001
From: Kiran Kumar K <kirankumark@marvell.com>
Date: Tue, 16 Mar 2021 09:42:27 +0530
Subject: [PATCH 09/13] octeontx2-af: Add support to parse cpt header with
 padding

commit 3aa9c2b03742fdac673bd75788b7c6bae827b8e7 from
git@git.assembla.com:cavium/WindRiver.linux.git

Packets received from CPT has data padding of 2 bytes for IPV6 packets
and 6 bytes for IPV4 packets, right after CPT header starting at byte 40.
So, while parsing the packet, ethertype has to be checked at byte 54 for
IPV6 packets and byte 58 for IPV4 packets. Adding generic KPU profile
changes to check for ethertype at byte 54 or 58 in case of packets
received from CPT.

Signed-off-by: Kiran Kumar K <kirankumark@marvell.com>
Change-Id: Ife4a00256e2aa925bdbba87f3bbd593760194c60
Reviewed-on: https://sj1git1.cavium.com/c/IP/SW/kernel/linux/+/48206
Tested-by: sa_ip-sw-jenkins <sa_ip-sw-jenkins@marvell.com>
Reviewed-by: Sunil Kovvuri Goutham <Sunil.Goutham@cavium.com>
Integrated-by: Abhishek Paliwal <paliwal.abhishek@windriver.com>
---
 .../marvell/octeontx2/af/npc_profile.h        | 173 +++---------------
 1 file changed, 27 insertions(+), 146 deletions(-)

diff --git a/drivers/net/ethernet/marvell/octeontx2/af/npc_profile.h b/drivers/net/ethernet/marvell/octeontx2/af/npc_profile.h
index 4059dd7c878e..82887e835a7e 100644
--- a/drivers/net/ethernet/marvell/octeontx2/af/npc_profile.h
+++ b/drivers/net/ethernet/marvell/octeontx2/af/npc_profile.h
@@ -1006,7 +1006,7 @@ static struct npc_kpu_profile_action ikpu_action_entries[] = {
 	},
 	{
 		NPC_ERRLEV_RE, NPC_EC_NOERR,
-		54, 58, 62, 0, 0,
+		40, 54, 58, 0, 0,
 		NPC_S_KPU1_CPT_HDR, 0, 0,
 		NPC_LID_LA, NPC_LT_NA,
 		0,
@@ -1884,61 +1884,25 @@ static struct npc_kpu_profile_cam kpu1_cam_entries[] = {
 	},
 	{
 		NPC_S_KPU1_CPT_HDR, 0xff,
-		NPC_ETYPE_IP,
-		0xffff,
-		0x0000,
-		0x0000,
-		0x0000,
 		0x0000,
-	},
-	{
-		NPC_S_KPU1_CPT_HDR, 0xff,
-		NPC_ETYPE_IP6,
 		0xffff,
-		0x0000,
-		0x0000,
-		0x0000,
-		0x0000,
-	},
-	{
-		NPC_S_KPU1_CPT_HDR, 0xff,
-		NPC_ETYPE_ARP,
+		NPC_ETYPE_IP,
 		0xffff,
 		0x0000,
 		0x0000,
-		0x0000,
-		0x0000,
 	},
 	{
 		NPC_S_KPU1_CPT_HDR, 0xff,
-		NPC_ETYPE_RARP,
-		0xffff,
-		0x0000,
-		0x0000,
-		0x0000,
 		0x0000,
-	},
-	{
-		NPC_S_KPU1_CPT_HDR, 0xff,
-		NPC_ETYPE_PTP,
 		0xffff,
-		0x0000,
-		0x0000,
-		0x0000,
-		0x0000,
-	},
-	{
-		NPC_S_KPU1_CPT_HDR, 0xff,
-		NPC_ETYPE_FCOE,
+		NPC_ETYPE_IP6,
 		0xffff,
 		0x0000,
 		0x0000,
-		0x0000,
-		0x0000,
 	},
 	{
 		NPC_S_KPU1_CPT_HDR, 0xff,
-		NPC_ETYPE_CTAG,
+		0x0000,
 		0xffff,
 		NPC_ETYPE_CTAG,
 		0xffff,
@@ -1947,75 +1911,48 @@ static struct npc_kpu_profile_cam kpu1_cam_entries[] = {
 	},
 	{
 		NPC_S_KPU1_CPT_HDR, 0xff,
-		NPC_ETYPE_CTAG,
-		0xffff,
 		0x0000,
-		0x0000,
-		0x0000,
-		0x0000,
-	},
-	{
-		NPC_S_KPU1_CPT_HDR, 0xff,
-		NPC_ETYPE_SBTAG,
 		0xffff,
-		0x0000,
-		0x0000,
-		0x0000,
-		0x0000,
-	},
-	{
-		NPC_S_KPU1_CPT_HDR, 0xff,
 		NPC_ETYPE_QINQ,
 		0xffff,
 		0x0000,
 		0x0000,
-		0x0000,
-		0x0000,
 	},
 	{
 		NPC_S_KPU1_CPT_HDR, 0xff,
-		NPC_ETYPE_ETAG,
-		0xffff,
-		0x0000,
-		0x0000,
 		0x0000,
-		0x0000,
-	},
-	{
-		NPC_S_KPU1_CPT_HDR, 0xff,
-		NPC_ETYPE_ITAG,
 		0xffff,
 		0x0000,
 		0x0000,
-		0x0000,
-		0x0000,
+		NPC_ETYPE_IP,
+		0xffff,
 	},
 	{
 		NPC_S_KPU1_CPT_HDR, 0xff,
-		NPC_ETYPE_MPLSU,
-		0xffff,
-		0x0000,
 		0x0000,
+		0xffff,
 		0x0000,
 		0x0000,
+		NPC_ETYPE_IP6,
+		0xffff,
 	},
 	{
 		NPC_S_KPU1_CPT_HDR, 0xff,
-		NPC_ETYPE_MPLSM,
-		0xffff,
-		0x0000,
 		0x0000,
+		0xffff,
 		0x0000,
 		0x0000,
+		NPC_ETYPE_CTAG,
+		0xffff,
 	},
 	{
 		NPC_S_KPU1_CPT_HDR, 0xff,
-		NPC_ETYPE_NSH,
-		0xffff,
-		0x0000,
 		0x0000,
+		0xffff,
 		0x0000,
 		0x0000,
+		NPC_ETYPE_QINQ,
+		0xffff,
 	},
 	{
 		NPC_S_KPU1_CPT_HDR, 0xff,
@@ -9676,46 +9613,6 @@ static struct npc_kpu_profile_action kpu1_action_entries[] = {
 		0,
 		0, 0, 0, 0,
 	},
-	{
-		NPC_ERRLEV_RE, NPC_EC_NOERR,
-		0, 0, 0, 3, 0,
-		NPC_S_KPU5_ARP, 56, 1,
-		NPC_LID_LA, NPC_LT_LA_CPT_HDR,
-		0,
-		0, 0, 0, 0,
-	},
-	{
-		NPC_ERRLEV_RE, NPC_EC_NOERR,
-		0, 0, 0, 3, 0,
-		NPC_S_KPU5_RARP, 56, 1,
-		NPC_LID_LA, NPC_LT_LA_CPT_HDR,
-		0,
-		0, 0, 0, 0,
-	},
-	{
-		NPC_ERRLEV_RE, NPC_EC_NOERR,
-		0, 0, 0, 3, 0,
-		NPC_S_KPU5_PTP, 56, 1,
-		NPC_LID_LA, NPC_LT_LA_CPT_HDR,
-		0,
-		0, 0, 0, 0,
-	},
-	{
-		NPC_ERRLEV_RE, NPC_EC_NOERR,
-		0, 0, 0, 3, 0,
-		NPC_S_KPU5_FCOE, 56, 1,
-		NPC_LID_LA, NPC_LT_LA_CPT_HDR,
-		0,
-		0, 0, 0, 0,
-	},
-	{
-		NPC_ERRLEV_RE, NPC_EC_NOERR,
-		8, 12, 0, 0, 0,
-		NPC_S_KPU2_CTAG2, 54, 1,
-		NPC_LID_LA, NPC_LT_LA_CPT_HDR,
-		NPC_F_LA_U_HAS_TAG | NPC_F_LA_L_WITH_VLAN,
-		0, 0, 0, 0,
-	},
 	{
 		NPC_ERRLEV_RE, NPC_EC_NOERR,
 		4, 8, 0, 0, 0,
@@ -9724,14 +9621,6 @@ static struct npc_kpu_profile_action kpu1_action_entries[] = {
 		NPC_F_LA_U_HAS_TAG | NPC_F_LA_L_WITH_VLAN,
 		0, 0, 0, 0,
 	},
-	{
-		NPC_ERRLEV_RE, NPC_EC_NOERR,
-		4, 8, 22, 0, 0,
-		NPC_S_KPU2_SBTAG, 54, 1,
-		NPC_LID_LA, NPC_LT_LA_CPT_HDR,
-		NPC_F_LA_U_HAS_TAG | NPC_F_LA_L_WITH_VLAN,
-		0, 0, 0, 0,
-	},
 	{
 		NPC_ERRLEV_RE, NPC_EC_NOERR,
 		4, 8, 0, 0, 0,
@@ -9742,42 +9631,34 @@ static struct npc_kpu_profile_action kpu1_action_entries[] = {
 	},
 	{
 		NPC_ERRLEV_RE, NPC_EC_NOERR,
-		8, 12, 26, 0, 0,
-		NPC_S_KPU2_ETAG, 54, 1,
-		NPC_LID_LA, NPC_LT_LA_CPT_HDR,
-		NPC_F_LA_U_HAS_TAG | NPC_F_LA_L_WITH_ETAG,
-		0, 0, 0, 0,
-	},
-	{
-		NPC_ERRLEV_RE, NPC_EC_NOERR,
-		18, 22, 26, 0, 0,
-		NPC_S_KPU2_ITAG, 54, 1,
+		8, 0, 6, 3, 0,
+		NPC_S_KPU5_CPT_IP, 60, 1,
 		NPC_LID_LA, NPC_LT_LA_CPT_HDR,
-		NPC_F_LA_U_HAS_TAG | NPC_F_LA_L_WITH_ITAG,
+		0,
 		0, 0, 0, 0,
 	},
 	{
 		NPC_ERRLEV_RE, NPC_EC_NOERR,
-		2, 6, 10, 2, 0,
-		NPC_S_KPU4_MPLS, 56, 1,
+		6, 0, 0, 3, 0,
+		NPC_S_KPU5_CPT_IP6, 60, 1,
 		NPC_LID_LA, NPC_LT_LA_CPT_HDR,
-		NPC_F_LA_L_WITH_MPLS,
+		0,
 		0, 0, 0, 0,
 	},
 	{
 		NPC_ERRLEV_RE, NPC_EC_NOERR,
-		2, 6, 10, 2, 0,
-		NPC_S_KPU4_MPLS, 56, 1,
+		4, 8, 0, 0, 0,
+		NPC_S_KPU2_CTAG, 58, 1,
 		NPC_LID_LA, NPC_LT_LA_CPT_HDR,
-		NPC_F_LA_L_WITH_MPLS,
+		NPC_F_LA_U_HAS_TAG | NPC_F_LA_L_WITH_VLAN,
 		0, 0, 0, 0,
 	},
 	{
 		NPC_ERRLEV_RE, NPC_EC_NOERR,
-		2, 0, 0, 2, 0,
-		NPC_S_KPU4_NSH, 56, 1,
+		4, 8, 0, 0, 0,
+		NPC_S_KPU2_QINQ, 58, 1,
 		NPC_LID_LA, NPC_LT_LA_CPT_HDR,
-		NPC_F_LA_L_WITH_NSH,
+		NPC_F_LA_U_HAS_TAG | NPC_F_LA_L_WITH_VLAN,
 		0, 0, 0, 0,
 	},
 	{
-- 
2.17.0

